File: 25-ReverseNodesInKGroup/25-ReverseNodesInKGroup.cpp
// Last updated: 12/6/2025, 5:55:00 am
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* reverseKGroup(ListNode* head, int k) {
        ListNode* dummy = new ListNode(0);
        ListNode* ret = dummy;
        ListNode* check = head;
        int num = 1;
        vector<int> last;
        while (check != NULL){
            if (num < k) {last.push_back(check->val); check = check->next;}
            else if (num == k){
                last.push_back(check->val);
                for (int i=last.size()-1; i>=0; i--){
                    ret->next = new ListNode(last[i]);
                    ret = ret->next;
                }
                check = check->next;
                num = 0;
                last.clear();
            }
            num++;
        }
       for (unsigned int i=0; i<last.size(); i++){
            ret->next = new ListNode(last[i]);
            ret = ret->next;
        }
        ret = dummy->next;
        delete dummy;
        delete check;
        return ret;
    }
};

File: 25-ReverseNodesInKGroup/25-ReverseNodesInKGroup.py
# Last updated: 12/6/2025, 5:54:54 am
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        if k == 1: return head
        mid = k//2
        convert = False
        def helper(node, index):
            nonlocal left, convert
            if index >= k or node == None: return node
            elif index == k-1: convert = True
            ret = helper(node.next, index+1)
            if convert and index >= mid:
                tmp = node.val
                node.val = left.val
                left.val = tmp
                left = left.next
            return ret
        
        node = head
        cnt = 0
        left = None
        while node:
            left = node
            node = helper(node, 0)
            convert = False
        return head